### PR TITLE
HPCC-13477 FLZ compressor fixes (2) for when input buffer > internal buffer

### PR DESCRIPTION
**Pull request by mckellyln**
This fix is also linked with HPCC-13491 and should address both issues.
1). Compressor write() needs to handle when buffer is larger than internal buffer.
2). If compressed output is same size as input then do not save compressed data.

Signed-off-by: Mark Kelly mark.kelly@lexisnexis.com

@jakesmith please review

### CODE REVIEW

**Comment by richardkchapman** [comment]
@jakesmith @ghalliday Please review

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

File: system/jlib/jflz.cpp [hunk 1]
```diff
@@ -52,6 +52,8 @@
 
 #include "jcrc.hpp"
 
+#include "jlog.hpp"
+
```
Comment (jakesmith) [inline system/jlib/jflz.cpp]: Doesn't look like this is needed..

File: system/jlib/jflz.cpp [hunk 2]
```diff
@@ -637,11 +640,13 @@ class jlib_decl CFastLZCompressor : public CInterface, public ICompressor
...
             return;
         size32_t toflush = (inlenblk==COMMITTED)?inlen:inlenblk;
+        if (toflush == 0)
+            return;
         assertex(outlen+sizeof(size32_t)*2+toflush+fastlzSlack(toflush)<=blksz);
         size32_t *cmpsize = (size32_t *)(outbuf+outlen);
         byte *out = (byte *)(cmpsize+1);
         *cmpsize = (size32_t)fastlz_compress(inbuf, (int)toflush, out, ht);
-        if (*cmpsize<=toflush) {
+        if (*cmpsize<toflush) {
```
Comment (jakesmith) [inline system/jlib/jflz.cpp]: >2). If compressed output is same size as input then do not save compressed data.

Is this related to here?
What were the consequences of doing so, inefficiency only?
Or was it ultimately the cause of the 'CMemoryBufferSerialStream::get read past end of stream error' errors ?

File: system/jlib/jflz.cpp [hunk 3]
```diff
@@ -676,6 +682,9 @@ class jlib_decl CFastLZCompressor : public CInterface, public ICompressor
 
     virtual void open(void *buf,size32_t max)
     {
+        wrmax = max / 4;
+        if (wrmax < 256)
+            wrmax = max;
```
Comment (jakesmith) [inline system/jlib/jflz.cpp]: Is this correct? e.g. if max = 1023.. max / 4 = 255, so wrmax = 1023 ?
Was it supposed to be something like wrmax = max>256?256:max on this line?

File: system/jlib/jflz.cpp (hunk 1 above)
Comment (mckellyln) [inline system/jlib/jflz.cpp]: I added it for some DBGLOG statements that I've since removed.  I will remove this include.

File: system/jlib/jflz.cpp (hunk 2 above)
Comment (mckellyln) [inline system/jlib/jflz.cpp]: Yes, this resolves the read past end of stream error for very small row.

File: system/jlib/jflz.cpp [hunk 4]
```diff
@@ -723,11 +732,28 @@ class jlib_decl CFastLZCompressor : public CInterface, public ICompressor
...
+        size32_t lenb = wrmax;
+        if (len > lenb)
+        {
+            byte *b = (byte *)buf;
+            size32_t lentot = 0;
+            while (lentot < len)
+            {
+                if ((lentot + lenb) > len)
+                    lenb = len - lentot;
+                size32_t amnt = write((const void *)(b+lentot), lenb);
```
Comment (jakesmith) [inline system/jlib/jflz.cpp]: to my mind, this isn't a natural use of recursion.
It will only ever recurse once, and fall out of into the block outside of this block as (len > lenb) will not be true.

It looks like this could be simplified as one while loop without a call, something like:
```code
while (len)
{
         if (len < lenb)
              lenb = len;
         if (lenb+inlen>inmax)
         {
             if (trailing)
                 return ... <amount written so far> (b-buf?);
             flushcommitted();
             if (lenb+inlen>inmax)
                  lenb = inmax-inlen;
         }
         memcpy(inbuf+inlen,b,len);
         b += lenb;
         inlen += lenb;
         len -= lenb;
}
```

File: system/jlib/jflz.cpp (hunk 3 above)
Comment (mckellyln) [inline system/jlib/jflz.cpp]: I want it to be one fourth of max but not too small, in which case it should be at most max.

File: system/jlib/jflz.cpp [hunk 5]
```diff
@@ -665,6 +670,7 @@ class jlib_decl CFastLZCompressor : public CInterface, public ICompressor
         outlen = 0;
         outbuf = NULL;      // only set on close
         bufalloc = 0;
+        wrmax = 0x10000;
```
Comment (jakesmith) [inline system/jlib/jflz.cpp]: Is there any point significance in setting to 0x10000?
As invalid / can't be used unless open() called, which sets..
Might be better to set to 0 and add comment that set by open.

File: system/jlib/jflz.cpp (hunk 3 above)
Comment (jakesmith) [inline system/jlib/jflz.cpp]: just seemed odd that:
if max is 1024, wrmax = 256
if max is 1023, wrmax = 1023

**Comment by jakesmith** [comment]
@mckellyln - a couple of questions

File: system/jlib/jflz.cpp (hunk 5 above)
Comment (mckellyln) [inline system/jlib/jflz.cpp]: ok, sure.

File: system/jlib/jflz.cpp (hunk 3 above)
Comment (mckellyln) [inline system/jlib/jflz.cpp]: wrmax = max / 4;
        if (wrmax < 256)
            wrmax = 256;
        if (wrmax < max)
            wrmax = max;

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

File: system/jlib/jflz.cpp (hunk 3 above)
Comment (mckellyln) [inline system/jlib/jflz.cpp]: I changed it so now I will try to write up to max bytes so no more divide by 4 with minimum

File: system/jlib/jflz.cpp (hunk 4 above)
Comment (mckellyln) [inline system/jlib/jflz.cpp]: I have re-written it to be a loop similar to above, trying to write up to wrmax (== max at open) bytes.  This also works well to keep small gaps when writing blocks out.

**Comment by mckellyln** [comment]
@jakesmith thank you for your review and comments.  I have made some changes and am running regression suite now.  Both example ecl tests (for 13477 and 13491) work ok.

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

**Comment by jakesmith** [comment]
@richardkchapman - looks good to merge

File: system/jlib/jflz.cpp [hunk 6]
```diff
@@ -723,16 +728,30 @@ class jlib_decl CFastLZCompressor : public CInterface, public ICompressor
...
+        byte *b = (byte *)buf;
+        size32_t written = 0;
+        while (len)
+        {
+            if (len < lenb)
+                lenb = len;
+            if (lenb+inlen>inmax) {
+                if (trailing)
+                    return written;
+                flushcommitted();
```
Comment (ghalliday) [inline system/jlib/jflz.cpp]: @mckellyln Wouldn't it be worth trying to fill inbuff up to the limit of inmax before compressing and flushing?

Comment (mckellyln) [inline system/jlib/jflz.cpp]: Yes, that is a good optimization.  I will work on that soon.
